packages/socket.io-promise/index.d.ts

```diff
@@ -0,0 +1,32 @@
+export interface IoPromiseAction {
+  type: string;
+}
+
+export interface IoPromiseSocket {
+  emit(event: string, ...args: any[]): any;
+}
+
+export interface RequestResponseRecord {
+  request: IoPromiseAction;
+  response: IoPromiseAction;
+}
+
+export type RequestResponsePair = [IoPromiseAction, IoPromiseAction];
+
+export type GetResponseRecord<RequestResponse extends RequestResponseRecord, Request extends RequestResponse['request']> =
+  Extract<RequestResponse, { request: Request, response: any }>['response'];
+export type GetResponsePair<RequestResponse extends RequestResponsePair, Request extends RequestResponse[0]> =
+  Extract<RequestResponse, [Request, any]>[1];
+
+export type IoPromiseRecord<RequestResponse extends RequestResponseRecord> = <Request extends RequestResponse['request']>(request: Request) => Promise<GetResponseRecord<RequestResponse, Request>>;
+export type IoPromisePair<RequestResponse extends RequestResponsePair> = <Request extends RequestResponse[0]>(request: Request) => Promise<GetResponsePair<RequestResponse, Request>>; 
+
+declare function createIoPromise<
+  RequestResponse extends RequestResponseRecord,
+>(socket: IoPromiseSocket, opts?: { eventName: string }): IoPromiseRecord<RequestResponse>;
+declare function createIoPromise<
+  RequestResponse extends RequestResponsePair,
+>(socket: IoPromiseSocket, opts?: { eventName: string }): IoPromisePair<RequestResponse>;
+declare function createIoPromise(socket: IoPromiseSocket, opts?: { eventName: string }): (request: IoPromiseAction) => Promise<IoPromiseAction>;
```

Comment:
Those overloads let you use generics to tell the middleware what types of request/response pairs exist for the app. Otherwise all you know is that a response has a `type` field and you have to do narrowing by explicitly checking that field's value in order to determine what the type of the response is, or by casting it. You can think of this as basically doing casting of the responses globally for the whole app, so you don't have to repeat yourself every time and with a bit more safety since it's automated for you.

```ts
interface MyRequest1 { ... }
interface MyResponse1 { ... }
interface MyRequest2 { ... }
interface MyResponse2 { ... }

const ioPromise = createIoPromise<[MyRequest1, MyResponse1] | [MyRequest2, MyResponse2]>(socket);
const myRequest1: MyRequest1 = ...;
const myRequest2: MyRequest2 = ...;
const myResponse1 = ioPromise(myRequest1); // infers type MyResponse1
const myResponse2 = ioPromise(myRequest2); // infers type MyResponse2
```